sable_history/src/pg_history_service.rs

```diff
@@ -0,0 +1,315 @@
+            Err(e) => {
+                tracing::error!("Could not get history channels: {e}");
+                return HashMap::new();
+            }
+            Ok(rows) => rows
+                .map(|row| -> Result<(TargetId, i64)> {
+                    let (channel_id, max_message_id): (i64, Uuid) = row?;
+                    let channel =
+                        TargetId::Channel(ChannelId::from(Snowflake::from(channel_id as u64)));
+
+                    let Some(ts) = max_message_id.get_timestamp() else {
+                        bail!("messages.id should be a UUID7, not {max_message_id}");
+                    };
+                    let (seconds, _) = ts.to_unix();
+                    let Ok(seconds) = seconds.try_into() else {
+                        bail!("message {max_message_id}'s UNIX timestamp is negative");
+                    };
+                    Ok((channel, seconds))
+                })
+                .try_collect()
+                .await
+                .unwrap_or_else(|e| {
+                    tracing::error!("Could not read rows: {e}");
+                    HashMap::new()
+                }),
+        }
+    }
+
+    async fn get_entries(
+        &self,
+        _user: UserId,
+        target: TargetId,
+        request: HistoryRequest,
+    ) -> Result<impl IntoIterator<Item = HistoricalEvent>, HistoryError> {
+        // TODO: access control
+        let TargetId::Channel(channel_id) = target else {
+            // TODO: PMs
+            return Err(HistoryError::InvalidTarget(target));
+        };
+
+        let mut connection_lock = self.database_connection.lock().await;
+
+        let db_channel_id = channel_id.as_u64() as i64;
+        let channel = match channels::dsl::channels
+            .find(db_channel_id)
+            .select(crate::models::Channel::as_select())
+            .first(&mut *connection_lock)
+            .await
+            .optional()
+        {
+            Ok(Some(channel)) => channel,
+            Ok(None) => return Err(HistoryError::InvalidTarget(target)),
+            Err(e) => {
+                tracing::error!("Could not check if channel exists: {e}");
+                return Err(HistoryError::InternalError(
+                    "Could not check if channel exists".to_string(),
+                ));
+            }
+        };
+
+        let base_query = messages::dsl::messages
+            .inner_join(historic_users::dsl::historic_users)
+            .select((
+                messages::dsl::id,
+                messages::dsl::timestamp,
+                messages::dsl::message_type,
+                messages::dsl::text,
+                historic_users::dsl::nick,
+                historic_users::dsl::ident,
+                historic_users::dsl::vhost,
+                historic_users::dsl::account_name,
+            ))
+            .filter(messages::dsl::target_channel.eq(db_channel_id));
+        Ok(match request {
+            HistoryRequest::Latest { to_ts, limit } => {
```

Comment:
Is there no commonality in these arms that can be broken out? They're quite hard to follow as they are.

Reply:
I can rewrite everything as a Between, I guess

Reply:
I managed to deduplicate most of the code: https://github.com/Libera-Chat/sable/pull/143/commits/a8f711b28921d0046392573fcd37af0ca139a4c4